numpy_dual_quaternion.c

```diff
@@ -0,0 +1,1436 @@
+    npy_intp i;                                                         \
+    for(i = 0; i < n; i++, ip1 += is1, ip2 += is2, op1 += os1) {        \
+      const arg_type1 in1 = *(arg_type1 *)ip1;                          \
+      const arg_type2 in2 = *(arg_type2 *)ip2;                          \
+      *((ret_type *)op1) = dual_quaternion_##func_name(in1, in2);            \
+    };                                                                  \
+  };
+// A couple special-case versions of the above
+#define BINARY_UFUNC(name, ret_type)                    \
+  BINARY_GEN_UFUNC(name, name, dual_quaternion, dual_quaternion, ret_type)
+#define BINARY_SCALAR_UFUNC(name, ret_type)                             \
+  BINARY_GEN_UFUNC(name##_scalar, name##_scalar, dual_quaternion, npy_double, ret_type) \
+  BINARY_GEN_UFUNC(scalar_##name, scalar_##name, npy_double, dual_quaternion, ret_type)
+// And these all do the work mentioned above, using the macros
+BINARY_UFUNC(add, dual_quaternion)
+BINARY_UFUNC(subtract, dual_quaternion)
+BINARY_UFUNC(multiply, dual_quaternion)
+BINARY_UFUNC(divide, dual_quaternion)
+BINARY_GEN_UFUNC(true_divide, divide, dual_quaternion, dual_quaternion, dual_quaternion)
+BINARY_GEN_UFUNC(floor_divide, divide, dual_quaternion, dual_quaternion, dual_quaternion)
+BINARY_UFUNC(power, dual_quaternion)
+BINARY_UFUNC(copysign, dual_quaternion)
+BINARY_UFUNC(equal, npy_bool)
+BINARY_UFUNC(not_equal, npy_bool)
+BINARY_UFUNC(less, npy_bool)
+BINARY_UFUNC(less_equal, npy_bool)
+BINARY_SCALAR_UFUNC(add, dual_quaternion)
+BINARY_SCALAR_UFUNC(subtract, dual_quaternion)
+BINARY_SCALAR_UFUNC(multiply, dual_quaternion)
+BINARY_SCALAR_UFUNC(divide, dual_quaternion)
+BINARY_GEN_UFUNC(true_divide_scalar, divide_scalar, dual_quaternion, npy_double, dual_quaternion)
+BINARY_GEN_UFUNC(floor_divide_scalar, divide_scalar, dual_quaternion, npy_double, dual_quaternion)
+BINARY_GEN_UFUNC(scalar_true_divide, scalar_divide, npy_double, dual_quaternion, dual_quaternion)
+BINARY_GEN_UFUNC(scalar_floor_divide, scalar_divide, npy_double, dual_quaternion, dual_quaternion)
+BINARY_SCALAR_UFUNC(power, dual_quaternion)
+
+
+#if PY_MAJOR_VERSION >= 3
+
+static struct PyModuleDef moduledef = {
+    PyModuleDef_HEAD_INIT,
+    "numpy_dual_quaternion",
+    NULL,
+    -1,
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+    NULL
+};
+
+#define INITERROR return NULL
+
+// This is the initialization function that does the setup
+PyMODINIT_FUNC PyInit_numpy_dual_quaternion(void) {
+
+#else
+
+#define INITERROR return
+
+// This is the initialization function that does the setup
+PyMODINIT_FUNC initnumpy_dual_quaternion(void) {
+
+#endif
+
+  PyObject *module;
+  PyObject *tmp_ufunc;
+  int dual_quaternionNum;
+  int arg_types[3];
+  PyObject* numpy;
+  PyObject* numpy_dict;
+
+  // Initialize a (for now, empty) module
+#if PY_MAJOR_VERSION >= 3
+  module = PyModule_Create(&moduledef);
+#else
+  module = Py_InitModule("numpy_dual_quaternion", DualQuaternionMethods);
```

Comment:
The identifier `DualQuaternionMethods` has never been declared, which results in not-compiling on my system. Have you been testing with python3 (which maybe does not have this problem)?

Reply:
Yeah I've been using python3 - I'll try testing with python2 now